kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala

```diff
@@ -1321,6 +1321,369 @@ class DeltaTableWritesSuite extends DeltaTableWriteSuiteBase with ParquetSuiteBa
+        .add("c", new ArrayType(StringType.STRING, false), true,
+          FieldMetadata.builder().putLong(ColumnMapping.COLUMN_MAPPING_ID_KEY, 2)
+            .putString(ColumnMapping.COLUMN_MAPPING_PHYSICAL_NAME_KEY, "c").build())
+
+      val e = intercept[IllegalArgumentException] {
+        table.createTransactionBuilder(engine, testEngineInfo, Operation.MANUAL_UPDATE)
+          .withSchema(engine, newSchema)
+          .build(engine).commit(engine, emptyIterable())
+      }
+
+      assert(e.getMessage.contains(
+        "Array field c must have exactly 1 nested ID"))
+    }
+  }
+
+  test("Test updating schema with adding an array and map type") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val schema = new StructType()
+        .add("a", IntegerType.INTEGER, true)
+
+      createTxn(engine, tablePath, isNewTable = true, schema, partCols = Seq.empty,
+        tableProperties = Map(TableConfig.COLUMN_MAPPING_MODE.getKey -> "id",
+          TableConfig.ICEBERG_COMPAT_V2_ENABLED.getKey -> "true")).commit(engine, emptyIterable())
+
+      val currentSchema = table.getLatestSnapshot(engine).getSchema
+      val newSchema = new StructType()
+        .add("a", IntegerType.INTEGER, true, currentSchema.get("a").getMetadata)
+        .add("arr", new ArrayType(StringType.STRING, false), true,
+          FieldMetadata.builder().putLong(ColumnMapping.COLUMN_MAPPING_ID_KEY, 2)
+            .putString(ColumnMapping.COLUMN_MAPPING_PHYSICAL_NAME_KEY, "arr")
+            .putFieldMetadata(ColumnMapping.COLUMN_MAPPING_NESTED_IDS_KEY,
+              FieldMetadata.builder().putLong("arr.element", 3).build()).build())
+        .add("map", new MapType(StringType.STRING, StringType.STRING, false), true,
+          FieldMetadata.builder().putLong(ColumnMapping.COLUMN_MAPPING_ID_KEY, 4)
+            .putString(ColumnMapping.COLUMN_MAPPING_PHYSICAL_NAME_KEY, "map")
+            .putFieldMetadata(ColumnMapping.COLUMN_MAPPING_NESTED_IDS_KEY,
+              FieldMetadata.builder().putLong("map.key", 5)
+                .putLong("map.value", 6).build()).build())
+
+        table.createTransactionBuilder(engine, testEngineInfo, Operation.MANUAL_UPDATE)
+          .withSchema(engine, newSchema)
+          .build(engine).commit(engine, emptyIterable())
+
+      val structType = table.getLatestSnapshot(engine).getSchema
+      assertColumnMapping(structType.get("a"), 1)
+      assertColumnMapping(structType.get("arr"), 2, "arr")
+      assertColumnMapping(structType.get("map"), 4, "map")
+      assert(structType.get("arr").getMetadata.get(ColumnMapping.COLUMN_MAPPING_NESTED_IDS_KEY)
+        == FieldMetadata.builder().putLong("arr.element", 3).build())
+      assert(structType.get("map").getMetadata.get(ColumnMapping.COLUMN_MAPPING_NESTED_IDS_KEY)
+        == FieldMetadata.builder().putLong("map.key", 5).putLong("map.value", 6).build())
+    }
+  }
+
```

Comment:
Add some tests with struct of struct, struct with map/array field, array of struct/map etc